### PR TITLE
Expand nested predicates

### LINKED ISSUE #10: Finish bug fix for unfold/fold statement removals
This could be a good task to onboard a newcomer to Scala

### LINKED ISSUE #20: Nested predicates aren't expanded
## Description

Original program:
```js
field left: Int
field right: Int
field middle: Int

predicate mid(this: Ref) {
  acc(this.middle, write)
}

predicate triple(this: Ref) {
  acc(this.left) && acc(this.right) && mid(this)
}

method tripleSum(this: Ref) returns (sum: Int)
  requires triple(this)
  ensures triple(this)
{
  unfold triple(this)
  unfold mid(this)
  sum := this.left + this.middle + this.right
  fold mid(this)
  fold triple(this)
}
```

Expected:
```js
method tripleSum(this: Ref) returns (sum: Int)
  requires acc(this.left) && acc(this.right, write) && acc(this.middle, write)
  ensures acc(this.left) && acc(this.right, write) && acc(this.middle, write)
{
  sum := this.left + this.middle + this.right
}
```

Actual:
```js
method tripleSum(this: Ref) returns (sum: Int)
  requires acc(this.left) && acc(this.right) && mid(this)
  ensures acc(this.left) && acc(this.right) && mid(this)
{
  unfold mid(this)
  sum := this.left + this.middle + this.right
  fold mid(this)
}
```

## Preliminary Investigation

This would be solved in `InlineRewrite.expandPredicates` by changing the traversal from `TopDown` to `BottomUp`, but we can't do that since the variable name context accumulates names from `QuantifiedExp`s outside-in, and changing it to bottom-up could break things.

### PR DESCRIPTION
**Pull request by ionathanch**
Fixes #20.

At first I had merged getting the predicate ids with expanding the predicates because it's really tricky getting all those nested names without doing the actual expansion, but since we only don't expand (mutually) recursive predicates, we don't need to return those names anyway.

Also, instead of passing in a set of predicate names that are or aren't expanded, I've abstracted that into a condition that's passed along (a Scala predicate for the Viper predicate names, so to speak) so that we can adjust the condition in `InlinePredicatePlugin` if ever needed.

Meanwhile, in `beforeVerify`, I've added getting *all* predicate ids in advance, filtered by whether those have bodies, since it's not really `expandPredicate`'s business to find out anyway; all it should do is check `cond`, check current scope, eat hot chip, and lie.

I've also commented out the stuff to do with the inline keyword. Once that's implemented, we can switch over, but I imagine we might also want to provide a global flag of some sort to say "please inline all predicates by default", then inline only those with the keyword in its absence? Then whether the cond uses `nonrecursivePredIds` or `nonrecursiveInlinePredIds` depends on the state of the flag. I suppose I should open a new stretch goal issue for that (but it could be as simple as adding a parser for a new type of declaration!).

Also it turns out I was working on the wrong branch so I've committed on top of a (fork of) [yoo/fix-fold-unfold-bug](/jyoo980/silver/tree/yoo/fix-fold-unfold-bug) oops, which I guess means this would also fix #10 if merged?

### CODE REVIEW

File: src/main/scala/viper/silver/plugin/standard/inline/InlinePredicatePlugin.scala [hunk 1]
```diff
@@ -27,21 +27,25 @@ class InlinePredicatePlugin extends SilverPlugin with ParserPluginTemplate
...
     val rewrittenMethods = input.methods.map { method =>
-      val inlinePredIds = input.extensions.collect({case InlinePredicate(p) => p.name}).toSet
-      val recursivePreds = checkRecursive(inlinePredIds, input) ++ checkMutualRecursive(inlinePredIds, input)
-      // TODO: Do we also need to inline in inhale/exhale/assert/assume and package/apply statements?
-      val (prePredIds, postPredIds) = getPrePostPredIds(method, input, inlinePredIds)
-      val prePredIdsNoRecur = prePredIds.diff(recursivePreds.map(_.name))
-      val postPredIdsNoRecur = postPredIds.diff(recursivePreds.map(_.name))
-      val inlinedPredMethod = inlinePredicates(method, input, prePredIdsNoRecur, postPredIdsNoRecur)
-      rewriteMethod(inlinedPredMethod, input, prePredIdsNoRecur, postPredIdsNoRecur)
+      val allPredIds = input.predicates.filter(_.body.isDefined).map(_.name).toSet
```
Comment (jyoo980) [inline src/main/scala/viper/silver/plugin/standard/inline/InlinePredicatePlugin.scala]: Just for the future, a sequence of `.filter` and `.map` can usually be represented by a `.collect`
```Scala
val allPredIds = input.predicates.collect {
    case pred if pred.body.isDefined => pred.name
}.toSet
```

File: src/main/scala/viper/silver/plugin/standard/inline/InlinePredicatePlugin.scala [hunk 2]
```diff
@@ -27,21 +27,25 @@ class InlinePredicatePlugin extends SilverPlugin with ParserPluginTemplate
...
+      val recursivePredIds = (checkRecursive(allPredIds, input) ++ checkMutualRecursive(allPredIds, input)).map(_.name)
```
Comment (jyoo980) [inline src/main/scala/viper/silver/plugin/standard/inline/InlinePredicatePlugin.scala:31]: This is a result of me poorly-designing the `InlineErrorChecker` API, not sure why I return a list of `Predicate` instead of just their names.

Lmk if fixing this is something that'll make your downstream tasks easier.

**Review by jyoo980 — APPROVED** [review]

**Comment by jyoo980** [comment]
> eat hot chip, and lie.

??

File: src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala [hunk 3]
```diff
@@ -3,114 +3,81 @@ package viper.silver.plugin.standard.inline
...
 
   /**
     * Removes given predicate unfolds and folds from statement.
     *
     * @param stmts A Seqn whose statements will be traversed
-    * @param predIds The set of predicates for which we will remove unfold and fold statements
+    * @param cond The condition a predicate must satisfy to no longer require (un)folding
     * @return The Seqn with all above unfolds and folds removed
     */
-  private[this] def removeFoldUnfolds(stmts: Seqn, predIds: Set[String]): Seqn = {
```
Comment (jyoo980) [inline src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala]: This is the fix @Alison-Li and I came up with, not sure if your changes still carry it through.

tl;dr we were just taking _all_ the pre/post condition ids, checking if any fold/unfold statements contained them, and removing them if true.

**Comment by ionathanch** [comment]
This has been tested with the following:

# Nested predicates

## Before
```js
field left: Int
field right: Int
field middle: Int

predicate p1(this: Ref) {
  acc(this.left) && p2(this)
}

predicate p2(this: Ref) {
  acc(this.middle) && p3(this)
}

predicate p3(this: Ref) {
  acc(this.right)
}

method sumTriple(this: Ref)
  returns  (sum: Int)
  requires p1(this)
  ensures  p1(this)
{
  unfold p1(this)
  unfold p2(this)
  unfold p3(this)
  sum := this.left + this.middle + this.right
  fold p3(this)
  fold p2(this)
  fold p1(this)
}
```

## After
```js
field left: Int
field right: Int
field middle: Int

predicate p1(this: Ref) {
  acc(this.left, write) && acc(p2(this), write)
}

predicate p2(this: Ref) {
  acc(this.middle, write) && acc(p3(this), write)
}

predicate p3(this: Ref) {
  acc(this.right, write)
}

method sumTriple(this: Ref) returns (sum: Int)
  requires acc(this.left, write) && (acc(this.middle, write) && acc(this.right, write))
  ensures acc(this.left, write) && (acc(this.middle, write) && acc(this.right, write))
{
  sum := this.left + this.middle + this.right
}

```

# Redundant folds/unfolds

## Before

```js
field content: Int

predicate boxed(box: Ref) {
  acc(box.content)
}

predicate bottom(box: Ref) { true }

method setBox(box: Ref, i: Int)
  requires boxed(box) && bottom(box)
  ensures  boxed(box)
  ensures  unfolding boxed(box) in box.content == i
{
  unfold boxed(box)
  unfold bottom(box)
  box.content := i
  fold boxed(box)
  unfold boxed(box)
  fold boxed(box)
}
```

## After

```js
field left: Int

field right: Int

field middle: Int

field content: Int

predicate boxed(box: Ref) {
  acc(box.content, write)
}

predicate bottom(box: Ref) { true }

method setBox(box: Ref, i: Int)
  requires acc(box.content, write) && true
  ensures acc(box.content, write)
  ensures box.content == i
{
  box.content := i
}
```

# Bonus

## Before
```js
method makeBox(i: Int)
  returns (box: Ref)
  // Note the missing `ensures boxed(box)`
{
  box := new(content)
  box.content := i
  fold boxed(box)
}
```

## After
```js
method makeBox(i: Int)
  returns (box: Ref)
{
  box := new(content)
  box.content := i
}
```

File: src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala (hunk 3 above)
Comment (jyoo980) [reply src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala]: Looks like the `pred` function should take care of it, since it's defined as
```Scala
val (prePredIds, postPredIds) = getPrePostPredIds(method, input, inlinePredIds)
val cond = { pred: String => nonrecursivePredIds(pred) }
```

**Comment by jyoo980** [comment]
Oh I get it: https://knowyourmeme.com/memes/eat-hot-chip-and-lie

👍

File: src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala (hunk 3 above)
Comment (ionathanch) [reply src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala]: I've made it more general: *all* fold/unfold statements will be removed as long as they're not (mutually) recursive. This takes care of cases where the pre-/postconditions don't actually specify the predicates, but they are being used, for instance when you fold up a predicate that isn't ensured in the postconditions. I'll add a test case for that lower down

File: src/main/scala/viper/silver/plugin/standard/inline/InlinePredicatePlugin.scala (hunk 1 above)
Comment (ionathanch) [reply src/main/scala/viper/silver/plugin/standard/inline/InlinePredicatePlugin.scala]: ooh neat, I'm gonna try that

**Review by alison-li — APPROVED** [review]
